src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -45,6 +45,61 @@ impl ScalarFunc {
         }
     }
 
+    #[inline]
+    pub fn oct_int<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        let mut num: u64 = 0;
+        let mut i: usize = 1;
+        let mut overflow = false;
+        let neg = match input[0] {
```

Comment:
What about use `self.children[0].eval_int(ctx, row)` to get input parameter and deal with overflow, like [implementation of builtin `bit_count`](https://github.com/tikv/tikv/blob/397b4a6e9ea6b5dd244247f726d031a4b53d059c/src/coprocessor/dag/expr/builtin_other.rs#L23) ?

Reply:
The `is_overflow`, Can diff the max_value overflow and min_value overflow?

Reply:
@yjhmelody  I take some extra experiments, and found that it seems that MySQL always treats input as unsigned. If so, there's no need to distinguish max/min value when overflow. 
The following are my experiments:
```sql
mysql> select oct(9223372036854775807);
+--------------------------+
| oct(9223372036854775807) |
+--------------------------+
| 777777777777777777777    |
+--------------------------+
1 row in set (0.00 sec)

mysql> select oct(-9223372036854775809);
+---------------------------+
| oct(-9223372036854775809) |
+---------------------------+
| 777777777777777777777     |
+---------------------------+
1 row in set (0.00 sec)

mysql> select oct(9223372036854775806);
+--------------------------+
| oct(9223372036854775806) |
+--------------------------+
| 777777777777777777776    |
+--------------------------+
1 row in set (0.00 sec)

mysql> select oct(-9223372036854775810);
+---------------------------+
| oct(-9223372036854775810) |
+---------------------------+
| 777777777777777777776     |
+---------------------------+
1 row in set (0.00 sec)

mysql> select oct(-9223372036854775808);
+---------------------------+
| oct(-9223372036854775808) |
+---------------------------+
| 1000000000000000000000    |
+---------------------------+
1 row in set (0.00 sec)

mysql> select oct(9223372036854775808);
+--------------------------+
| oct(9223372036854775808) |
+--------------------------+
| 1000000000000000000000   |
+--------------------------+
1 row in set (0.01 sec)

mysql> select oct(-11111111111111111111111111111111);
+----------------------------------------+
| oct(-11111111111111111111111111111111) |
+----------------------------------------+
| 1777777777777777777777                 |
+----------------------------------------+
1 row in set (0.00 sec)

mysql> select oct(11111111111111111111111111111111);
+---------------------------------------+
| oct(11111111111111111111111111111111) |
+---------------------------------------+
| 1777777777777777777777                |
+---------------------------------------+
1 row in set (0.00 sec)

```

I'll check the MySQL source code to confirm this. You can take a look at this as well. By the way, shall we add these cases into tests?

Reply:
I have a question. Should I rewrite the tests just as `builtin bit_count ` ?  I found it has different Datum type to test.  While I only test the Datum::Bytes because of the `oct(11111111111111111111111111111111)`
Let me check your meaning, if the number overflow, we consider it 64 binary numbers all 1.  And, when we parse it to oct, we only map it binary value to a determined oct value.
If so, it seem that these are only little work need to do.

Reply:
1. Maybe write the tests just as `builtin bit_count` is better because this is builtin `OctInt`. Current form of tests is more suitable for `OctString`, I think. 
2. We'd better check the MySQL source code first. I'm not sure now, as well

Reply:
@yjhmelody Hi, I've checked MySQL source code. in [sql/item_strfunc.cc](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/sql/item_strfunc.cc#L3970), after getting an `ulonglong` input, the following line is executed to transfer `ulonglong` into `string`:
```c++
     ptr= longlong2str(dec, ans, to_base))
```

`longlong2str` is defined in [include/m_string.h](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/include/m_string.h#L222) and implemented in [strings/longlong2str.c](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/strings/longlong2str.c#L50).
you can see in [line 63](https://github.com/mysql/mysql-server/blob/913071c0b16cc03e703308250d795bc381627e37/strings/longlong2str.c#L63) that there's a conversion for negative inputs:
```c
      /* Avoid integer overflow in (-val) for LLONG_MIN (BUG#31799). */
      uval = (ulonglong)0 - uval;
```

Reply:
So, what's the conclusion?
I was originally based FormatUint which used in tidb and is similar to the definition which you showed me.
I'm a little confused about the version of C.

Reply:
TiDB make `OctInt` simple because big integer-literal are treated as string constant and is processed in `octString`.   I think we can refine the current implementation a little, convert it into `OctString`, so we can make `OctInt` more simple as your original version.